pkg/controller/eks/addon/setup.go

```diff
@@ -0,0 +1,232 @@
+	"github.com/aws/aws-sdk-go/service/eks/eksiface"
+	"github.com/pkg/errors"
+	"k8s.io/client-go/util/workqueue"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+
+	xpv1 "github.com/crossplane/crossplane-runtime/apis/common/v1"
+	"github.com/crossplane/crossplane-runtime/pkg/event"
+	"github.com/crossplane/crossplane-runtime/pkg/logging"
+	"github.com/crossplane/crossplane-runtime/pkg/meta"
+	"github.com/crossplane/crossplane-runtime/pkg/ratelimiter"
+	"github.com/crossplane/crossplane-runtime/pkg/reconciler/managed"
+	"github.com/crossplane/crossplane-runtime/pkg/resource"
+
+	"github.com/crossplane/provider-aws/apis/eks/v1alpha1"
+	awsclients "github.com/crossplane/provider-aws/pkg/clients"
+)
+
+const (
+	errNotEKSCluster    = "managed resource is not an EKS cluster custom resource"
+	errKubeUpdateFailed = "cannot update EKS cluster custom resource"
+	errListTags         = "cannot list tags"
+	errTagResource      = "cannot tag resource"
+	errUntagResource    = "cannot untag resource"
+)
+
+// SetupAddon adds a controller that reconciles Clusters.
+func SetupAddon(mgr ctrl.Manager, l logging.Logger, rl workqueue.RateLimiter, poll time.Duration) error {
+	name := managed.ControllerName(v1alpha1.AddonGroupKind)
+	opts := []option{
+		setupHooks,
+	}
+
+	return ctrl.NewControllerManagedBy(mgr).
+		Named(name).
+		WithOptions(controller.Options{
+			RateLimiter: ratelimiter.NewController(rl),
+		}).
+		For(&v1alpha1.Addon{}).
+		Complete(managed.NewReconciler(mgr,
+			resource.ManagedKind(v1alpha1.AddonGroupVersionKind),
+			managed.WithExternalConnecter(&connector{kube: mgr.GetClient(), opts: opts}),
+			managed.WithInitializers(managed.NewDefaultProviderConfig(mgr.GetClient()), &tagger{kube: mgr.GetClient()}),
+			managed.WithReferenceResolver(managed.NewAPISimpleReferenceResolver(mgr.GetClient())),
+			managed.WithPollInterval(poll),
+			managed.WithLogger(l.WithValues("controller", name)),
+			managed.WithRecorder(event.NewAPIRecorder(mgr.GetEventRecorderFor(name)))))
+}
+
+func setupHooks(e *external) {
+	e.preObserve = preObserve
+	e.postObserve = postObserve
+	e.lateInitialize = lateInitialize
+	h := &hooks{client: e.client, kube: e.kube}
+	e.isUpToDate = h.isUpToDate
+	e.preUpdate = preUpdate
+	e.postUpdate = h.postUpdate
+	e.preCreate = preCreate
+	e.postCreate = postCreate
+	e.preDelete = preDelete
+}
+
+type hooks struct {
+	client eksiface.EKSAPI
+	kube   client.Client
+}
+
+func preObserve(_ context.Context, cr *v1alpha1.Addon, obj *awseks.DescribeAddonInput) error {
+	obj.ClusterName = cr.Spec.ForProvider.ClusterName
+	return nil
+}
+
+func postObserve(_ context.Context, cr *v1alpha1.Addon, _ *awseks.DescribeAddonOutput, obs managed.ExternalObservation, err error) (managed.ExternalObservation, error) {
+	if err != nil {
+		return managed.ExternalObservation{}, err
+	}
+
+	cr.SetConditions(xpv1.Available())
```

Comment:
is this the only status we have for AddOns ?

Reply:
Fixed: Added additional conditions for addon status